executor/cluster_reader.go

```diff
@@ -215,3 +222,106 @@ func (e *clusterConfigRetriever) retrieve(ctx sessionctx.Context) ([][]types.Dat
 	}
 	return finalRows, nil
 }
+
+type clusterServerInfoRetriever struct {
+	extractor    *plannercore.ClusterTableExtractor
+	serverInfoTP diagnosticspb.ServerInfoType
+}
+
+// retrieve implements the clusterRetriever interface
+func (e *clusterServerInfoRetriever) retrieve(ctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.extractor.SkipRequest {
+		return nil, nil
+	}
+	return e.dataForClusterInfo(ctx, e.serverInfoTP)
+}
+
+func (e *clusterServerInfoRetriever) dataForClusterInfo(ctx sessionctx.Context, infoTp diagnosticspb.ServerInfoType) ([][]types.Datum, error) {
+	serversInfo, err := infoschema.GetClusterServerInfo(ctx)
+	if err != nil {
+		return nil, err
+	}
+	ipMap := make(map[string]struct{}, len(serversInfo))
+	rows := make([][]types.Datum, 0, len(serversInfo)*10)
+	for _, srv := range serversInfo {
+		// Skip some node type which has been filtered in WHERE cluase
+		// e.g: SELECT * FROM cluster_config WHERE type='tikv'
+		if len(e.extractor.NodeTypes) > 0 && !e.extractor.NodeTypes.Exist(srv.ServerType) {
+			continue
+		}
+		address := srv.Address
+		// Skip some node address which has been filtered in WHERE cluase
+		// e.g: SELECT * FROM cluster_config WHERE address='192.16.8.12:2379'
+		if len(e.extractor.Addresses) > 0 && !e.extractor.Addresses.Exist(address) {
+			continue
+		}
+		if srv.ServerType == "tidb" {
+			address = srv.StatusAddr
+		}
+		ip := srv.StatusAddr
+		if idx := strings.Index(address, ":"); idx != -1 {
+			ip = address[:idx]
+		}
+		if _, ok := ipMap[ip]; ok {
+			continue
+		}
+		ipMap[ip] = struct{}{}
+		items, err := getServerInfoByGRPC(address, infoTp)
```

Comment:
Better to call `getServerInfoByGRPC` parallel, which can improve performance if the cluster is huge.